Libraries/PeriphDrivers/Source/SPI/spi_ai87_v2.c

```diff
@@ -0,0 +1,920 @@
+            if (error != E_NO_ERROR) {
+                return error;
+            }
+
+            break;
+
+        // Target Select 2 (TS2 - L. SS2 pin)
+        case 2:
+            error = MXC_GPIO_Config(&gpio_cfg_spi0_ts2);
+            if (error != E_NO_ERROR) {
+                return error;
+            }
+
+            error = MXC_GPIO_SetVSSEL(gpio_cfg_spi0_ts2.port, vssel, gpio_cfg_spi0_ts2.mask);
+            if (error != E_NO_ERROR) {
+                return error;
+            }
+
+            break;
+
+        default:
+            return E_BAD_PARAM;
+        }
+#endif
+    }
+
+    return E_NO_ERROR;
+}
+
+int MXC_SPI_SetFrequency(mxc_spi_regs_t *spi, unsigned int hz)
+{
+    return MXC_SPI_RevA2_SetFrequency((mxc_spi_reva_regs_t *)spi, hz);
+}
+
+unsigned int MXC_SPI_GetFrequency(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_GetFrequency((mxc_spi_reva_regs_t *)spi);
+}
+
+int MXC_SPI_SetFrameSize(mxc_spi_regs_t *spi, int frame_size)
+{
+    return MXC_SPI_RevA2_SetFrameSize((mxc_spi_reva_regs_t *)spi, frame_size);
+}
+
+int MXC_SPI_GetFrameSize(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_GetFrameSize((mxc_spi_reva_regs_t *)spi);
+}
+
+int MXC_SPI_SetModeIF(mxc_spi_regs_t *spi, mxc_spi_interface_t mode)
+{
+    return MXC_SPI_RevA2_SetModeIF((mxc_spi_reva_regs_t *)spi, mode);
+}
+
+mxc_spi_interface_t MXC_SPI_GetModeIF(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_GetModeIF((mxc_spi_reva_regs_t *)spi);
+}
+
+int MXC_SPI_SetClkMode(mxc_spi_regs_t *spi, mxc_spi_clkmode_t clk_mode)
+{
+    return MXC_SPI_RevA2_SetClkMode((mxc_spi_reva_regs_t *)spi, clk_mode);
+}
+
+mxc_spi_clkmode_t MXC_SPI_GetClkMode(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_GetClkMode((mxc_spi_reva_regs_t *)spi);
+}
+
+int MXC_SPI_SetRegisterCallback(mxc_spi_regs_t *spi, mxc_spi_callback_t callback, void *data)
+{
+    return MXC_SPI_RevA2_SetRegisterCallback((mxc_spi_reva_regs_t *)spi, callback, data);
+}
+
+int MXC_SPI_GetActive(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_GetActive((mxc_spi_reva_regs_t *)spi);
+}
+
+/* ** DMA-Specific Functions ** */
+
+int MXC_SPI_DMA_Init(mxc_spi_init_t *init)
+{
+    return MXC_SPI_RevA2_DMA_Init(init);
+}
+
+bool MXC_SPI_DMA_GetInitialized(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_DMA_GetInitialized((mxc_spi_reva_regs_t *)spi);
+}
+
+int MXC_SPI_DMA_GetTXChannel(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_DMA_GetTXChannel((mxc_spi_reva_regs_t *)spi);
+}
+
+int MXC_SPI_DMA_GetRXChannel(mxc_spi_regs_t *spi)
+{
+    return MXC_SPI_RevA2_DMA_GetRXChannel((mxc_spi_reva_regs_t *)spi);
+}
+
+int MXC_SPI_DMA_SetRequestSelect(mxc_spi_req_t *req)
+{
+    int spi_num;
+    int tx_reqsel = -1;
+    int rx_reqsel = -1;
+
+    spi_num = MXC_SPI_GET_IDX(req->spi);
+    if (spi_num < 0 || spi_num >= MXC_SPI_INSTANCES) {
+        return E_INVALID;
+    }
+
+    if (req->tx_buffer != NULL) {
+        switch (spi_num) {
+        case 0:
```

Comment:
This is prevalent in the ME17, ME18, AI85, and AI87. Originally, I believe this numbering flip was made to support the previous design where the AHB SPI had a lower instance number than the APB SPI, which caused problems with the RISCV seeing incorrect SPI instances (RISCV only sees the APB SPI). We usually order our SPI instances with APB SPI instances first (lower instance numbers) and AHB instances second (higher instance numbers).

The SPI instance ordering might have been updated with later designs, but the current instance numbering was kept since it is a breaking change.